### PR TITLE
NMS-8238: Topology providers should provide a description for display

### PR DESCRIPTION
**Pull request by christianpape**
* JIRA: http://issues.opennms.org/browse/NMS-8238
* Bamboo: http://bamboo.internal.opennms.com:8085/browse/OPENNMS-ONMS609

### CODE REVIEW

File: features/topology-map/plugins/org.opennms.features.topology.plugins.topo.application/src/main/resources/OSGI-INF/blueprint/blueprint.xml [hunk 1]
```diff
@@ -17,6 +17,9 @@ xsi:schemaLocation="http://www.osgi.org/xmlns/blueprint/v1.0.0 http://www.osgi.o
 	<!-- Make it hop aware -->
 	<bean id="applicationTopologyHopProvider" class="org.opennms.features.topology.api.support.VertexHopGraphProvider">
 		<argument ref="applicationTopologyProvider" />
+
+		<property name="name" value="Application Topology Provider"/>
+		<property name="description" value="This Topology Provider displays the defined application topologies."/>
```
Comment (mvrueden) [inline features/topology-map/plugins/org.opennms.features.topology.plugins.topo.application/src/main/resources/OSGI-INF/blueprint/blueprint.xml]: Maybe `This Topology Provider displays all defined Applications and its calculated state`

**Comment by mvrueden** [comment]
What do you think about introducing a "MetaInfo" class/interface which the providers expose instead of adding more methods to the TopologyProvider itself.

**Comment by pbrane** [comment]
> On Apr 12, 2016, at 5:41 AM, mvrueden <notifications@github.com> wrote:
> 
> What do you think about introducing a "MetaInfo" class/interface which the providers expose instead of adding more methods to the TopologyProvider itself.

I like it.

:David

**Comment by mvrueden** [comment]
We can also try annotations.

File: features/topology-map/plugins/org.opennms.features.topology.plugins.topo.bsm/src/main/resources/OSGI-INF/blueprint/blueprint.xml [hunk 2]
```diff
@@ -15,9 +15,11 @@
         <argument ref="transactionAwareBeanProxyFactory" />
         <property name="businessServiceManager" ref="businessServiceManager" />
     </bean>
-    <bean id="bsmTopologyHopProvider"
-        class="org.opennms.features.topology.api.support.VertexHopGraphProvider">
+    <bean id="bsmTopologyHopProvider" class="org.opennms.features.topology.api.support.VertexHopGraphProvider">
         <argument ref="bsmTopologyProvider" />
+
+        <property name="name" value="Business Service Monitoring Topology Provider"/>
```
Comment (jeffgdotorg) [inline features/topology-map/plugins/org.opennms.features.topology.plugins.topo.bsm/src/main/resources/OSGI-INF/blueprint/blueprint.xml]: `BSM Topology Provider`

File: features/topology-map/plugins/org.opennms.features.topology.plugins.topo.bsm/src/main/resources/OSGI-INF/blueprint/blueprint.xml [hunk 3]
```diff
@@ -15,9 +15,11 @@
...
+        <property name="description" value="This Topology Provider displays the hierarchy of the defined Business Services and their computed operational state."/>
```
Comment (jeffgdotorg) [inline features/topology-map/plugins/org.opennms.features.topology.plugins.topo.bsm/src/main/resources/OSGI-INF/blueprint/blueprint.xml]: This Topology Provider displays the hierarchy of the defined Business Services and their computed operational states.

File: features/topology-map/plugins/org.opennms.features.topology.plugins.topo.linkd/src/main/resources/OSGI-INF/blueprint/blueprint.xml [hunk 4]
```diff
@@ -68,6 +68,9 @@ xsi:schemaLocation="http://www.osgi.org/xmlns/blueprint/v1.0.0
     <bean id="topologyACLProvider" class="org.opennms.features.topology.plugins.topo.linkd.internal.NodeACLVertexProvider">
         <argument ref="topologyHopProvider" />
         <argument ref="nodeDao" />
+
+        <property name="name" value="Enhanced Linkd Topology Provider"/>
+        <property name="description" value="This Topology Provider displays the topology information discovered by the Enhanced Linkd daemon. It uses the SNMP information of several protocols like OSPF, ISIS, LLDP and CDP to generate an overall topology."/>
```
Comment (jeffgdotorg) [inline features/topology-map/plugins/org.opennms.features.topology.plugins.topo.linkd/src/main/resources/OSGI-INF/blueprint/blueprint.xml]: Sounds good already :)

File: features/topology-map/plugins/org.opennms.features.topology.plugins.topo.vmware/src/main/resources/OSGI-INF/blueprint/blueprint.xml [hunk 5]
```diff
@@ -11,6 +11,9 @@
           class="org.opennms.features.topology.plugins.topo.vmware.internal.VmwareTopologyProvider">
         <property name="nodeDao" ref="nodeDao"/>
         <property name="ipInterfaceDao" ref="ipInterfaceDao"/>
+
+        <property name="name" value="VMware Topology Provider"/>
+        <property name="description" value="The VMware Topology Provider displays the infrastructure information gathered by the VMware Provisioning process."/>
```
Comment (jeffgdotorg) [inline features/topology-map/plugins/org.opennms.features.topology.plugins.topo.vmware/src/main/resources/OSGI-INF/blueprint/blueprint.xml]: Sounds good

**Comment by christianpape** [comment]
I changed the code to use MetaInfo objects for name and description. The annotation approach was not possible since the OSGI proxies did not have them anymore.

File: features/topology-map/org.opennms.features.topology.app/src/main/java/org/opennms/features/topology/app/internal/TopologyUI.java [hunk 6]
```diff
@@ -369,11 +371,47 @@ public String getTitle(GraphContainer container) {
...
+                return container.getSelectionManager().getSelectedEdgeRefs().isEmpty()
+                        && container.getSelectionManager().getSelectedVertexRefs().isEmpty();
+            }
+
+            @Override
+            public String getTitle(GraphContainer container) {
+                MetaInfo metaInfo = getGraphContainer().getBaseTopology().getMetaInfo();
+
+                if (Objects.isNull(metaInfo)) {
+                    metaInfo = new DefaultMetaInfo();
```
Comment (mvrueden) [inline features/topology-map/org.opennms.features.topology.app/src/main/java/org/opennms/features/topology/app/internal/TopologyUI.java]: Can we encapsulate this in a `getMetaInfo(container)` method?

File: features/topology-map/plugins/org.opennms.features.topology.plugins.topo.linkd/src/main/java/org/opennms/features/topology/plugins/topo/linkd/internal/EnhancedLinkdTopologyProvider.java [hunk 7]
```diff
@@ -53,7 +53,9 @@
 import org.opennms.features.topology.api.topo.AbstractSearchProvider;
 import org.opennms.features.topology.api.topo.AbstractVertex;
 import org.opennms.features.topology.api.topo.Criteria;
+import org.opennms.features.topology.api.topo.DefaultMetaInfo;
 import org.opennms.features.topology.api.topo.Edge;
+import org.opennms.features.topology.api.topo.MetaInfo;
```
Comment (mvrueden) [inline features/topology-map/plugins/org.opennms.features.topology.plugins.topo.linkd/src/main/java/org/opennms/features/topology/plugins/topo/linkd/internal/EnhancedLinkdTopologyProvider.java]: Do we need these imports? It is defined in AbstractTopologyProvider and not used in this class.

**Comment by mvrueden** [comment]
Perfect! :+1: 
Let's wait for bamboo before we merge.

**Pull request merged**